### PR TITLE
Add method to await Crystal jobs

### LINKED ISSUE #50: [FEATURE] Add get_lifecycle_stage method to CrystalConnector
**Is your feature request related to a problem? Please describe.**
We currently lack a way to effectively monitor the lifecycle of a crystal job through the CrystalConnector class. This hinders our ability to launch a crystal job and wait for its completion.

**Describe the solution you'd like**
We'd like to introduce lifecycle monitoring functionality in the CrystalConnector class. This will allow us to launch a crystal job and track its status, which can be used for e.g. awaiting completion. [Similar functionality is available in the BeastConnector](https://github.com/SneaksAndData/esd-services-api-client/blob/cd6d35a1b2d324cd8743722f8c99397e32575dd3/esd_services_api_client/beast/_connector.py#L229).

**Additional context**
Having the ability to monitor the lifecycle of the CrystalConnector class will improve our ability to manage crystal jobs, track their progress, and react to any issues that may arise during their execution. This enhancement will ultimately lead to more efficient and streamlined job management in our application. Additionally, this feature will be particularly useful when launching crystal jobs from Airflow, as it will enable better orchestration and monitoring of the workflows, ensuring smooth execution and minimizing potential issues.

### PR DESCRIPTION
**Pull request by matt035343**
Fixes #50

### CODE REVIEW

File: esd_services_api_client/crystal/_connector.py [hunk 1]
```diff
@@ -280,3 +287,47 @@ def dispose(self) -> None:
...
+        Await for a list of submitted Crystal jobs to complete.
+
+        :param algorithm: Name of an algorithm.
+        :param run_ids: Request identifiers assigned to the jobs by Crystal.
+        :param tags: Request tags assigned to the jobs by a client.
+        """
+        return self._await_runs(algorithm=algorithm, run_ids=run_ids, tags=tags)
+
+    def _await_runs(self, algorithm: str, run_ids: Optional[List[str]] = None, tags: Optional[List[str]] = None) -> Iterator[RequestResult]:
+        if tags is None:
```
Comment (george-zubrienko) [inline esd_services_api_client/crystal/_connector.py:304]: logic for tags here

File: esd_services_api_client/crystal/_connector.py [hunk 2]
```diff
@@ -280,3 +287,47 @@ def dispose(self) -> None:
...
+            tags = []
+
+        if run_ids is None:
```
Comment (george-zubrienko) [inline esd_services_api_client/crystal/_connector.py:307]: run_id logic here

File: esd_services_api_client/crystal/_connector.py [hunk 3]
```diff
@@ -280,3 +287,47 @@ def dispose(self) -> None:
...
+        ]
+
+        while len(tags) > 0:
+            for tag in tags[:]:
+                results = self._retrieve_runs(tag=tag, algorithm=algorithm)
+                if all(result.status in finished_statuses for result in results):
+                    tags.remove(tag)
+                    yield from results
+            if len(tags) > 0:
+                doze(1)
```
Comment (george-zubrienko) [inline esd_services_api_client/crystal/_connector.py:324]: should doze a bit more than 1s due to performance issues with that endpoint

File: esd_services_api_client/crystal/_connector.py [hunk 4]
```diff
@@ -280,3 +287,47 @@ def dispose(self) -> None:
...
+
+        while len(run_ids) > 0:
```
Comment (george-zubrienko) [inline esd_services_api_client/crystal/_connector.py:326]: if you do this in a thread, you won't need for loop or while, i think

**Review by george-zubrienko — APPROVED** [review]

File: esd_services_api_client/crystal/_connector.py [hunk 5]
```diff
@@ -280,3 +287,47 @@ def dispose(self) -> None:
         Gracefully dispose object.
         """
         self.http.close()
+
+    def await_runs(
+        self, algorithm: str, run_ids: Optional[List[str]] = None, tags: Optional[List[str]] = None
```
[the remark this answers is not in the document]
Comment (matt035343) [reply esd_services_api_client/crystal/_connector.py]: Done

File: esd_services_api_client/crystal/_connector.py (hunk 4 above)
Comment (matt035343) [reply esd_services_api_client/crystal/_connector.py]: Threaded but still looped

File: esd_services_api_client/crystal/_connector.py (hunk 1 above)
Comment (matt035343) [reply esd_services_api_client/crystal/_connector.py]: no longer needed

**Review by george-zubrienko — APPROVED** [review]
huge :)

**Comment by matt035343** [comment]
create_package

**Pull request merged**